src/prefect/context.py

```diff
@@ -610,32 +620,32 @@ def emit_events(self, state: State) -> None:
         else:
             return
 
-        asset_deps_related: list[Asset] = []
+        # If we have no downstream assets, this not a materialization
+        if not self.downstream_assets:
+            return
 
-        # Emit reference events for direct asset dependencies
-        for asset in self.direct_asset_dependencies:
+        # Emit reference events for all upstream assets (direct + inherited)
+        all_upstream_assets = self.upstream_assets + self.direct_asset_dependencies
```

Comment:
is there any risk of duplicates here?

Reply:
Good call, I'm going to convert all of these to sets